### ISSUE TITLE
NNBD migration suggests to start with `leaves` instead of `root`

### ISSUE THREAD

**Issue by tomaszpolanski** [dart-lang/site-www#3451]
### Page URL

https://dart.dev/null-safety/migration-guide#step1-wait

### Describe the problem

In the explanation on what dependency to migrate first, it is said that users should migrate `leaves` first:
```
We strongly recommend migrating code in order, with the leaves of the dependency graph being migrated first.
```

### Expected fix

`leaves` should be changed to `root`:

```
We strongly recommend migrating code in order, with the root of the dependency graph being migrated first.
```

`root` has 0 parents and 1+ children
`leave` has 1 parent and 0 children.

The explanation below is valid as it suggests first migrate A which is a `root`

**Comment by parlough** [comment]
Hey @tomaszpolanski, thanks for opening an issue!

I may be missing something, but I think you want to migrate the leaf packages with no children(dependencies) as you don't need to wait for any children to be migrated yet for that package to properly support sound null safety. Then you can migrate the parents and so on, until you reach the root, in this case C which depends on the previous ones.

Where do you see it suggest to migrate C first? I found this quote, which suggests to migrate C last.

> then A should be migrated to null safety first, then B, then C.

**Comment by tomaszpolanski** [comment]
Hey @parlough,
thanks for the response. 
First, you are right A should be first migrated and not C - that is my typo, will amend it.

Second, here is an
![graph](https://user-images.githubusercontent.com/1532227/128413192-b9602323-34ec-463d-a286-bfe6ee6267d2.png)
 image of what I understand as root package/dependency and leaves:

I have migrated a couple of packages and I know that first, you should migrate the root package that has no dependencies OR all dependencies are already migrated - otherwise compiler/analyzer will complain.

The confusion is the definition of root - my image shows how it looks in trees (like a binary tree) - maybe just in dependency terminology, the definitions are reverted.

**Comment by parlough** [comment]
Hmm, I guess in this example it was trying to describe C as the root package, as the one you're trying to migrate.

For example, in this visualization of the pubviz dependencies: https://kevmoo.github.io/pubviz/

Say I want to migrate pubviz to null safety, I see it as the root package, then I have to look at its leaf nodes and update those first. 

I'll pull in @kwalrath as she worked with others to write this piece and will likely have some more context and perhaps ideas on how to improve the clarity of this section.

**Comment by kwalrath** [comment]
It's funny, but people think of roots and leaves in different ways, maybe because graphs tend to show the root at the top and the leaves at the bottom, which is counterintuitive if you think about real-life trees.

I believe the description is correct, as written, but we can always make it clearer. I'm just not sure how.

**Comment by tomaszpolanski** [comment]
I agree that the description is correct - the only controversial issue is `root` vs `leaves`. 
Another way to explaining this is upstream vs downstream dependency but it might also bring some confusion.

The reason why I've created this issue is that already twice I had to explain it to confused collegues.

**Comment by kwalrath** [comment]
> The reason why I've created this issue is that already twice I had to explain it to confused colleagues.

Thanks, it really helps to have that context. We could emphasize whether your dependencies are null safe, and mention but deemphasize _root/leaf_ and related terms.

**Comment by antfitch** [comment]
Hi folks, I've just looked at the graph and A appears to be a leaf of C. So the wording looks correct to me. I haven't seen any other issues around this section, so closing this.
